### PR TITLE
Moved the nightshade implementation into a feature.

### PR DESCRIPTION
**Pull request by pelson**
## Rationale

Creates a Nightshade feature that can be added to a cartopy GeoAxes, rather than a ``.nightshade`` method on the GeoAxes. (follow up to #1135)
The reason is one of cleanliness on the GeoAxes API, as well as a better abstraction for future cartopy growth (e.g. alternative backends)

## Implications

* ``cartopy/feature.py`` moved to ``cartopy/feature/__init__.py``
* ``cartopy/mpl/_solar.py`` inlined into ``cartopy/feature/nightshade.py`` (solar physics don't need to belong in a matplotlib interface)
* adding a nightshade to a GeoAxes now looks like:

```
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
from cartopy.feature.nightshade import Nightshade


ax = plt.axes(projection=ccrs.PlateCarree())
ax.add_feature(Nightshade())
plt.show()
```

### CODE REVIEW

File: lib/cartopy/examples/nightshade.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import matplotlib.pyplot as plt
+import cartopy.crs as ccrs
+from cartopy.feature.nightshade import Nightshade
+
+
+ax = plt.axes(projection=ccrs.PlateCarree())
```
Comment (QuLogic) [inline lib/cartopy/examples/nightshade.py]: Need to update the existing example, then.

Comment (pelson) [reply lib/cartopy/examples/nightshade.py]: Happy to remove mine. Hadn't actually checked if one existed, my apologies - was using this example as a quick proof-of-concept.

Comment (QuLogic) [reply lib/cartopy/examples/nightshade.py]: It's a part of the Aurora example.

File: lib/cartopy/feature/nightshade.py [hunk 2]
```diff
@@ -16,7 +16,95 @@
...
+        """
+        Shade the darkside of the Earth, accounting for refraction.
+
+        Parameters
+        ----------
+        date
+            A UTC datetime object used to calculate the position of the sun.
+            Default: datetime.datetime.utcnow()
+        delta
+            Float (degrees), stepsize to determine the resolution of the
```
Comment (QuLogic) [inline lib/cartopy/feature/nightshade.py]: Can you move the `float` to the type instead?

File: lib/cartopy/feature/nightshade.py [hunk 3]
```diff
@@ -16,7 +16,95 @@
...
+        x[npts:] = 180 - omega0[npts:]
+
+        kwargs.setdefault('facecolor', color)
+        kwargs.setdefault('alpha', alpha)
+
+        geom = sgeom.Polygon(zip(x, y))
+        return super().__init__([geom], rotated_pole, **kwargs)
 
 
 def julian_day(date):
```
Comment (QuLogic) [inline lib/cartopy/feature/nightshade.py]: Do we want these to be public though?

**Comment by dopplershift** [comment]
👍 to making this a feature.

**Comment by pelson** [comment]
OK. Ready for final review I think.

File: lib/cartopy/examples/nightshade.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+date = datetime.datetime(1999, 12, 31, 12)
+
+plt.title('Night time shading for {}'.format(date))
```
Comment (QuLogic) [inline lib/cartopy/examples/nightshade.py]: `ax.set_title`

Comment (pelson) [reply lib/cartopy/examples/nightshade.py]: OK I accept your cleanliness 😜

Comment (pelson) [reply lib/cartopy/examples/nightshade.py]: I also *attempted* to improve the description.

**Pull request merged**